src/main.c

```diff
@@ -522,15 +533,41 @@ static void async_nb_run_sync(neverbleed_iobuf_t *buf, void (*transaction_cb)(ne
 
 static void async_nb_read_ready(h2o_socket_t *sock, const char *err)
 {
-    struct async_nb_transaction_t *transaction = async_nb_pop(&async_nb.read_queue);
-    assert(transaction != NULL);
+    fd_set rfds;
+    struct timeval tv;
+    int ret;
+    FD_ZERO(&rfds);
+    FD_SET(neverbleed_get_fd(neverbleed), &rfds);
+
+    // neverbleed will never write half a response because we limit the number of in-flight transactions with NEVERBLEED_MAX_IN_FLIGHT_TX
+    // read responses until the neverbleed fd is no longer read ready
+    while (async_nb.read_queue.len > 0) {
+        struct async_nb_transaction_t *transaction = async_nb_pop(&async_nb.read_queue);
+        assert(transaction != NULL);
 
-    async_nb_run_sync(transaction->buf, neverbleed_transaction_read);
+        async_nb_run_sync(transaction->buf, neverbleed_transaction_read);
+        transaction->on_read_complete(transaction);
+
+        tv.tv_sec = 0;
+        tv.tv_usec = 0;
+        while ((ret = select(neverbleed_get_fd(neverbleed) + 1, &rfds, NULL, NULL, &tv)) == -1 && (errno == EAGAIN || errno == EINTR))
+            ;
+        if (ret == -1)
+            h2o_fatal("select(2):%d\n", errno);
+
+        if (!ret)
+            break;
+    }
+
+    // resume writing if there's room
+    if (async_nb.read_queue.len < NEVERBLEED_MAX_IN_FLIGHT_TX) {
+        if (!h2o_timer_is_linked(&async_nb.write_pending_timer)) {
+            h2o_timer_link(h2o_socket_get_loop(async_nb.sock), 0, &async_nb.write_pending_timer);
```

Comment:
I suppose we could just call `async_nb_submit_write_pending` here directly, my thought process was to yield to the event loop instead and the callback will be called after it's finished processing. Thoughts?

Reply:
It is entirely fine to write to the same socket from the read callback; so unless there is a reason to delay, we do not need a timer.

If the concern is neverbleed doing too much work due to us writing earlier, I think it would take time for neverbleed to process the requests and therefore that there is no difference between writing here or at the end of the event loop.

WDYT?